src/jit.c

```diff
@@ -0,0 +1,1557 @@
+    emit1(state, 0xff);
+    emit1(state, 0xe6);
+
+    /* Epilogue */
+    state->exit_loc = state->offset;
+
+    /* Move register 0 into rax */
+    if (map_register(VM_REG_0) != RAX)
+        emit_mov(state, map_register(VM_REG_0), RAX);
+
+    /* Deallocate stack space by restoring RSP from JIT R10. */
+    emit_mov(state, map_register(VM_REG_10), RSP);
+
+    if (!(ARRAYS_SIZE(nonvolatile_reg) % 2))
+        emit_alu64_imm32(state, 0x81, 0, RSP, 0x8);
+
+    /* Restore platform non-volatile registers */
+    for (uint32_t i = 0; i < ARRAYS_SIZE(nonvolatile_reg); i++)
+        emit_pop(state, nonvolatile_reg[ARRAYS_SIZE(nonvolatile_reg) - i - 1]);
+
+    /* Return */
+    emit1(state, 0xc3);
+#elif defined(__aarch64__)
+    uint32_t register_space = ARRAYS_SIZE(callee_reg) * 8 + 2 * 8;
+    state->stack_size = align_up(STACK_SIZE + register_space, 16);
+    emit_addsub_immediate(state, true, AS_SUB, SP, SP, state->stack_size);
+
+    /* Set up frame */
+    emit_loadstorepair_immediate(state, LSP_STPX, R29, R30, SP, 0);
+    /* In ARM64 calling convention, R29 is the frame pointer. */
+    emit_addsub_immediate(state, true, AS_ADD, R29, SP, 0);
+
+    /* Save callee saved registers */
+    for (size_t i = 0; i < ARRAYS_SIZE(callee_reg); i += 2) {
+        emit_loadstorepair_immediate(state, LSP_STPX, callee_reg[i],
+                                     callee_reg[i + 1], SP, (i + 2) * 8);
+    }
+
+    emit_uncond_branch_reg(state, BR_BR, R1);
+    /* Epilogue */
+    state->exit_loc = state->offset;
+
+    /* Move register 0 into R0 */
+    if (map_register(0) != R0) {
+        emit_logical_register(state, true, LOG_ORR, R0, RZ, map_register(0));
+    }
+
+    /* Restore callee-saved registers).  */
+    for (size_t i = 0; i < ARRAYS_SIZE(callee_reg); i += 2) {
+        emit_loadstorepair_immediate(state, LSP_LDPX, callee_reg[i],
+                                     callee_reg[i + 1], SP, (i + 2) * 8);
+    }
+    emit_loadstorepair_immediate(state, LSP_LDPX, R29, R30, SP, 0);
+    emit_addsub_immediate(state, true, AS_ADD, SP, SP, state->stack_size);
+    emit_uncond_branch_reg(state, BR_RET, R30);
+#endif
+}
+
+#define GEN(inst, code)                                                       \
+    static void do_##inst(struct jit_state *state UNUSED, riscv_t *rv UNUSED, \
+                          rv_insn_t *ir UNUSED)                               \
+    {                                                                         \
+        code;                                                                 \
+    }
+#include "rv32_jit.c"
+#undef GEN
+
+static void do_fuse1(struct jit_state *state, riscv_t *rv UNUSED, rv_insn_t *ir)
+{
+    opcode_fuse_t *fuse = ir->fuse;
+    for (int i = 0; i < ir->imm2; i++) {
+        emit_load_imm(state, temp_reg[0], fuse[i].imm);
+        emit_store(state, S32, temp_reg[0], parameter_reg[0],
+                   offsetof(struct riscv_internal, X) + 4 * fuse[i].rd);
```

Comment:
Why not using `offsetof(riscv_t, X)`?

Since `offsetof(struct riscv_internal, X)` plus a certain offset is frequently used, we can introduce a macro to simplify the specification of memory locations.